### PR TITLE
GMG: fix max/min of the evaluated viscosities

### PR DESCRIPTION
**Pull request by zjiaqi2018**
Fix max/min of the evaluated viscosities in `evaluate_material_model()`. @tjhei

### CODE REVIEW

**Review by tjhei — APPROVED** [review]
Perfect, thank you.

**Comment by tjhei** [comment]
Please apply the test changes...

File: source/simulator/stokes_matrix_free.cc [hunk 1]
```diff
@@ -1908,6 +1911,8 @@ namespace aspect
         const double imbalance = MGTools::workload_imbalance(sim.triangulation);
         sim.pcout << "    GMG workload imbalance: " << imbalance << std::endl
                   << "    Stokes solver: " << std::flush;
+
+        sim.pcout << "Viscosity range: " << viscosity_min << " - " << viscosity_max << std::endl;
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1915]: Move this new code above the line 1913. Right now you are printing in the middle of the line of the Stokes iterations.

**Review by tjhei — APPROVED** [review]

**Review by tjhei — APPROVED** [review]
awesome, thank you.

File: include/aspect/stokes_matrix_free.h [hunk 2]
```diff
@@ -573,6 +573,11 @@ namespace aspect
        */
       bool do_timings;
 
+      /**
+       * The max/min of the evaluated viscosities.
+       */
+      double viscosity_max, viscosity_min;
```
Comment (gassmoeller) [inline include/aspect/stokes_matrix_free.h:579]:
```suggestion
      double minimum_viscosity;
      double maximum_viscosity;
```

File: source/simulator/stokes_matrix_free.cc [hunk 3]
```diff
@@ -1370,8 +1370,8 @@ namespace aspect
 
     const QGauss<dim> quadrature_formula (sim.parameters.stokes_velocity_degree+1);
 
-    double min_el = std::numeric_limits<double>::max();
-    double max_el = std::numeric_limits<double>::lowest();
+    double min_el_local = std::numeric_limits<double>::max();
+    double max_el_local = std::numeric_limits<double>::lowest();
```
Comment (gassmoeller) [inline source/simulator/stokes_matrix_free.cc:1374]:
```suggestion
    double minimum_viscosity_local = std::numeric_limits<double>::max();
    double maximum_viscosity_local = std::numeric_limits<double>::lowest();
```

File: source/simulator/stokes_matrix_free.cc [hunk 4]
```diff
@@ -1492,7 +1495,7 @@ namespace aspect
                   // of the evaluated viscosity on the active level.
                   for (unsigned int q=0; q<n_q_points; ++q)
                     active_cell_data.viscosity(cell, q)[i]
-                      = std::min(std::max(values_on_quad[q], min_el), max_el);
+                      = std::min(std::max(values_on_quad[q], viscosity_min), viscosity_max);
```
Comment (gassmoeller) [inline source/simulator/stokes_matrix_free.cc:1498]: Was this the bug before? Did we limit the viscosity only to the local minimum and maximum? But I do not understand why this had caused the issue, limiting to the local minimum and maximum should only be stricter than limiting to the global minimum/maximum. Can you explain what the problem was?

**Review by gassmoeller — COMMENTED** [review]
I have a suggestion for renaming variables to better conform to other places in ASPECT (not your fault, the naming was a bit weird before already). Also can you explain what you fixed? Just so that if someone (e.g. I) read over this PR again in the future know what happened. Also please add a changelog entry if it was a fixed bug.

**Comment by tjhei** [comment]
> Was this the bug before? Did we limit the viscosity only to the local minimum and maximum? But I do not understand why this had caused the issue, limiting to the local minimum and maximum should only be stricter than limiting to the global minimum/maximum. Can you explain what the problem was?

Without the MPI communication we could have slightly wrong viscosities on coarser GMG levels maybe resulting in higher iteration numbers (if at all). This is not a fix for #4466 but one step along the way. That will come as a followup.

**Comment by tjhei** [comment]
Not sure if we need a changelog entry at this point.

**Comment by gassmoeller** [comment]
Thanks for the explanation. No if this is not a fix for https://github.com/geodynamics/aspect/issues/4466 we dont need a changelog yet. Do you want to do the renaming as part of this PR or the next? I am also fine pushing that to the next Pr and just merge this one.

**Comment by tjhei** [comment]
I would prefer if Jiaqi would do the rename here. @zjiaqi2018 ?

**Pull request merged**